src/Microsoft.SqlTools.Hosting.v2/Utility/Logger.cs

```diff
@@ -93,207 +159,269 @@ public static Logger Instance
+        {
+        }
 
-            if (logWriter != null)
+        // No unmanaged resources in this class, and it is sealed.
+        // No finalizer needed. See http://stackoverflow.com/a/3882819/613130
+        public void Dispose()
+        {
+            if (IsValueCreated)
             {
-                logWriter.Write(
-                    logLevel,
-                    logMessage,
-                    callerName,
-                    callerSourceFile,
-                    callerLineNumber);
+                Value.Dispose();
             }
         }
     }
 
-    internal class LogWriter : IDisposable
+    /// <summary>
+    /// This listener has the same behavior as TextWriterTraceListener except it controls how the 
+    /// options: TraceOptions.DateTime, TraceOptions.ProcessId and TraceOptions.ThreadId is written to the output stream.
+    /// This listener writes the above options, if turned on, inline with the message 
+    /// instead of writing them to indented fields as is the case with TextWriterTraceListener.
+    /// This implementation also lazily initializes the underlying tracelistener
+    /// </summary>
+    /// <remarks>
+    /// Implementation of this is a lazily initialize trace listener that is partly inspired 
+    /// by: https://stackoverflow.com/questions/30664527/how-to-stop-streamwriter-to-not-to-create-file-if-nothing-to-write
+    /// </remarks>
+    internal sealed class SqlToolsTraceListener : TraceListener
     {
-        private object logLock = new object();
-        private TextWriter textWriter;
-        private LogLevel minimumLogLevel = LogLevel.Verbose;
+        Lazy<TextWriterTraceListener> _lazyListerner;
+        private TextWriterTraceListener Listener => _lazyListerner.Value;
+
+        public SqlToolsTraceListener(string file, string listenerName = "") : base(listenerName) 
+            // Wrapping around lazy to make sure that we do not create file if the log.Write events are getting filtered out. i.e. the log file actually gets created the first time an actual write to log file happens.
+            => _lazyListerner = new Lazy<TextWriterTraceListener>(
+                valueFactory: () => new TextWriterTraceListener(new StreamWriter(file, append: true), listenerName),
+                // LazyThreadSafetyMode.PublicationOnly mode ensures that we keep trying to create the listener (especially the file that write) on all future log.write events even if previous attempt(s) have failed
+                mode: LazyThreadSafetyMode.PublicationOnly 
+             );
+        #region forward actual write events to the underlying listener.
+        public override void Write(string message) => Listener.Write(message);
+
+        public override void WriteLine(string message) => Listener.WriteLine(message);
+
+        /// <Summary> 
+        /// Closes the <see cref="System.Diagnostics.TextWriterTraceListener.Writer"> so that it no longer 
+        ///    receives tracing or debugging output.</see>
```

Comment:
normally I thought it would be a self-close `<see cref="System.Diagnostics.TextWriterTraceListener.Writer" />`. How does this doc look? Do we lost the contents `so that it no longer 
        ///    receives tracing or debugging output` here?

Reply:
yes if we call the close and the application continue to send log messages then they would be lost and not sent to the listeners. However we can reconfigure new listeners by setting the logfile name.